test/lib/jdk/test/lib/util/ForceGC.java

```diff
@@ -23,68 +23,65 @@
 
 package jdk.test.lib.util;
 
-import java.lang.ref.Cleaner;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.TimeUnit;
+import java.lang.ref.PhantomReference;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
 import java.util.function.BooleanSupplier;
 
 /**
  * Utility class to invoke System.gc()
  */
 public class ForceGC {
-    private final static Cleaner cleaner = Cleaner.create();
-
-    private final CountDownLatch cleanerInvoked;
-    private Object o;
-    private int gcCount = 0;
-
-    public ForceGC() {
-        this.o = new Object();
-        this.cleanerInvoked = new CountDownLatch(1);
-        cleaner.register(o, cleanerInvoked::countDown);
-    }
-
-    private void doit(int iter) {
-        try {
-            for (int i = 0; i < 10; i++) {
-                System.gc();
-                gcCount++;
-                if (cleanerInvoked.await(100L, TimeUnit.MILLISECONDS)) {
-                    return;
-                }
-            }
-        } catch (InterruptedException unexpected) {
-            throw new AssertionError("unexpected InterruptedException");
-        }
+    /**
+     * Causes the current thread to wait until the {@code booleanSupplier}
+     * returns true, or a predefined waiting time (10 seconds) elapses.
+     *
+     * @param booleanSupplier boolean supplier
+     * @return true if the {@code booleanSupplier} returns true, or false
+     *     if did not complete after 10 Seconds
+     */
+    public static boolean wait(BooleanSupplier booleanSupplier) {
+        return wait(booleanSupplier, 10000L);
```

Comment:
For the max waiting time, instead of a fixed value I strongly recommend choosing some base waiting time, say 1 second, and scaling it by the jtreg timeout factor.

Reply:
I also have similar feedback.  The timeout factor can be found in the `test.timeout.factor` system property, if present.    When running tests with -Xcomp in a fastdebug build, the timeout factor is currently configured to 10.   Setting the base waiting time to 1 second seems reasonable.

Reply:
Good point!  I changed to use 1 second as the base waiting time.  I was not very sure if 1 second is big enough for the GC cleanup, but the testing result on my laptop and VMs, which are super faster, looks good to me.  Please help to check mach5 result.